Improve troubleshooting docs (#21137)
MaterializeInc/materialize: doc/user/content/manage/troubleshooting.md

        AND mdod.id = mda.id
        -- exclude regions and just return operators
        AND mda.address NOT IN (
            SELECT DISTINCT address[:list_length(address) - 1]
            FROM mz_internal.mz_dataflow_addresses
        )

)

SELECT *
FROM histograms
WHERE duration > '100 millisecond'::interval
ORDER BY duration DESC
```
```
 id  |                 name                 |             dataflow_name              | count |    duration
-----+--------------------------------------+----------------------------------------+-------+-----------------
 408 | persist_source::decode_and_mfp(u510) | Dataflow: materialize.qck.num_bids     |     1 | 00:00:01.073741
 408 | persist_source::decode_and_mfp(u510) | Dataflow: materialize.qck.num_bids     |     1 | 00:00:00.53687
 374 | persist_source::decode_and_mfp(u509) | Dataflow: materialize.qck.num_bids     |     1 | 00:00:00.268435
 408 | persist_source::decode_and_mfp(u510) | Dataflow: materialize.qck.num_bids     |     2 | 00:00:00.268435
 429 | FormArrangementKey                   | Dataflow: materialize.qck.num_bids     |     2 | 00:00:00.134217
 (5 rows)
```

Note that this relation contains a lot of information.

> The query therefore filters all duration below `100 millisecond`.

Reviewer: ```suggestion
The query therefore filters all durations below `100 millisecond`.
```

  - /ops/troubleshooting/
---

This page describes several SQL queries you can run to diagnose performance issues with Materialize.

## Mental model and basic terminology

> When you create a materialized view and issue a query, Materialize creates a so-called dataflow. A dataflow consists of instructions on how to respond to data input and to changes to that data. Once executed, the dataflow computes the result of the SQL query, polls the source for updates, and then incrementally updates the query results when new data arrives.

Reviewer 2: This is imprecise. Dataflows can be created in three cases:

1. When creating a materialized view
2. When creating an index
3. When running a one-off (aka ad-hoc) query, but only if that query does not use the "fast path" execution strategy (i.e., looking at an existing index)

I would also avoid using the word "poll" as the system is not poll-based. I would prefer "waits for updates from the sources" or similar.

CREATE INDEX num_bids_idx ON num_bids (item);
```

The query of the materialized view joins the relations `bids` and `auctions`, groups by `auctions.item` and determines the number of bids per auction. To understand how this SQL query is translated to a dataflow, we can use `EXPLAIN` to display the plan used to evaluate the join.
Note that you can also explain the plan of queries and common views with `EXPLAIN` and `EXPLAIN VIEW`, respectively.

```sql
EXPLAIN MATERIALIZED VIEW num_bids
```
```
                Optimized Plan
-----------------------------------------------
 materialize.public.num_bids:                 +
   Reduce group_by=[#0] aggregates=[count(*)] +
     Project (#3)                             +
       Filter (#1 < #4)                       +
         Join on=(#0 = #2) type=differential  +
           ArrangeBy keys=[[#0]]              +
             Project (#2, #4)                 +
               Get materialize.public.bids    +
           ArrangeBy keys=[[#0]]              +
             Project (#0, #2, #3)             +
               Get materialize.public.auctions+

(1 row)
```

The plan describes the specific operators that are used to evaluate the query.

> Some of these operators resemble relational algebra or map reduce style operators (`Filter`, `Join on`, `Project`).

Author: Oh, that makes so much more sense!

Others are specific to Materialize (`Get`, `ArrangeBy`).

In general, a high level understanding of what these operators do is sufficient for effective debugging:
`Filter` filters records, `Join on` joins records from two or more inputs, `Map` applies a function to transform records, etc.

> The `ArrangeBy` operator indexes data and stores it in memory for fast point lookups.

Reviewer 3: `ArrangeBy` is a bit tricky, because in many cases it is just signaling that an existing arrangement will be used. So if there is an `ArrangeBy` in the plan, it might or might not build an arrangement as part of this dataflow.

Reviewer 3: So, I'd suggest to just skip trying to explain `ArrangeBy` for now.

Btw. I have some plans to make EXPLAIN print an `ArrangeBy` differently based on whether it's reading an existing index or creating a new arrangement. So later we can clarify this.

Author: You are right, talking about the nuances may be too much. I've removed the discussion for now. The PR looks interesting!

SELECT
    mdo.id,
    mdo.name,
    mse.elapsed_ns / 1000 * '1 MICROSECONDS'::interval AS elapsed_time
FROM mz_internal.mz_scheduling_elapsed AS mse,
    mz_internal.mz_dataflow_operators AS mdo,
    mz_internal.mz_dataflow_addresses AS mda
WHERE mse.id = mdo.id AND mdo.id = mda.id AND list_length(address) = 1
ORDER BY elapsed_ns DESC
```
```
 id  |                  name                  |  elapsed_time
-----+----------------------------------------+-----------------
 354 | Dataflow: materialize.qck.num_bids     | 02:05:25.756836
 578 | Dataflow: materialize.qck.num_bids_idx | 00:15:04.838741
 (2 rows)
```

These results show that Materialize spend the most time keeping the materialized view `num_bids` up to date.
Followed by the work on the index `avg_bids_idx` that has been defined on the materialized view.

### Finding expensive operators within a dataflow

The previous query is a good starting point to get an overview of the work happening in the cluster because it only returns dataflows.

The `mz_scheduling_elapsed` relation also contains details for regions and operators.
Removing the condition `list_length(address) = 1` will include the regions and operators in the result.

> But be aware that every row shows aggregated times for all the element it contains.

Reviewer: ```suggestion
But be aware that every row shows aggregated times for all the elements it contains.
```

The `elapsed_time` reported for the dataflows above also includes the elapsed time for all the regions and operators they contain.

> But because the parent child relationship is not always obvious,

Reviewer: ```suggestion
But because the parent-child relationship is not always obvious,
```

the results containing a mixture of dataflows, regions, and operators can be a bit hard to interpret.
The following query therefore only returns operators from the  `mz_scheduling_elapsed` relation.
You can further drill down by adding a filter condition that matches the name of a specific dataflow.

```sql
-- Extract raw elapsed time information for operators
WITH leaves AS (
    SELECT *
    FROM mz_internal.mz_dataflow_addresses AS outer
    WHERE
        outer.address NOT IN (
            SELECT inner.address[:list_length(outer.address)]
            FROM mz_internal.mz_dataflow_addresses AS inner
            WHERE inner.id != outer.id

>         )

Author: Hmm, this seems to work and is indeed easier to understand (and avoids at least one join). Thanks!
```
SELECT
    mdod.id,
    mdod.name,
    mdod.dataflow_name,
    mse.elapsed_ns / 1000 * '1 MICROSECONDS'::interval AS elapsed_time
FROM mz_internal.mz_scheduling_elapsed AS mse,
    mz_internal.mz_dataflow_addresses AS mda,
    mz_internal.mz_dataflow_operator_dataflows AS mdod
WHERE
    mse.id = mdod.id AND mdod.id = mda.id
    -- exclude all operators that are regions
    AND mda.address NOT IN (
        SELECT DISTINCT address[:list_length(address) - 1]
        FROM mz_internal.mz_dataflow_addresses
    )
ORDER BY elapsed_ns DESC
```

  - /ops/troubleshooting/
---

This page describes several SQL queries you can run to diagnose performance issues with Materialize.

## Mental model and basic terminology

When you create a materialized view and issue a query, Materialize creates a dataflow. A dataflow consists of instructions on how to respond to data input and to changes to that data. Once executed, the dataflow computes the result of the SQL query, polls the source for updates, and then incrementally updates the query results when new data arrives.

> Materialize dataflows act on collections of data. A collection provides a data stream of updates as they happen. To provide fast access to the changes to individual records, the collection can be represented in an alternate form, indexed on data to present the sequence of changes the collection has undergone. This indexed representation is called an arrangement.

Reviewer 4: Link to the arrangements doc on the word arrangement

Reviewer 4: I think this paragraph has too much detail.

“To provide fast access to changing records, the collection can be stored in memory in an index structure called an arrangement.”

something like that?

> ### Translating SQL to Differential Dataflow

Author: For now I've removed references to differential dataflow, because it does not add too much value. It may make sense to add it back later (e.g. if we wanted to explain how certain patterns are translated like left joins & certain group by). But right now, I"m not feeling strongly about it.

> ### Translating SQL to Differential Dataflow

Reviewer 5: I almost feel we can get really specific in dd details. I don't want to call it out by name so users don't feel they must learn two tools.

> To make these concepts a bit more tangible, let's look at the example from the getting started guide.

Author: That's a fair point. If we ever change the getting started guide, it'll break that part of the docs. The reason I've done it this way is because there are several queries that show up in the examples diagnostics queries. But it' probably even less confusing, if there is only one view doing work instead of several. I'll make it self contained.

(1 row)
```

The plan describes the specific Differential Dataflow operators that are used to evaluate the query.
Some of these operators resemble relational algebra or map reduce style operators (`Filter`, `Join on`, `Map`). 
Others are specific to differential dataflow (`Get`, `ArrangeBy`).

In general, a high level understanding of what these operators do is sufficient for effective debugging:
`Filter` filters records, `Join on` joins records from two or more inputs, `Map` applies a function to transform records, etc.
The `ArrangeBy` operator indexes data and stores it in memory for fast point lookups.
It's not important to have a deep understanding of all these operators for effective debugging, though.

Behind the scenes, the operator graph is turned into a dataflow that can be evaluated by the Differential Dataflow runtime.
The dataflow is organized in a hierarchical structure. 
It contains operators and so-called regions, which in turn contain operators and (sub)regions.

In our example, the dataflow contains a InputRegion and a BuildRegion.
And the BuildRegion contains the operators from the above plan.

> ![Regions and operator visualization](/images/regions-and-operators.png)

Reviewer 6: If we're using this image, we should clearly tell users how to access the memory visualizer.

Again, it's not too important for our purposes to understand what these regions do and how they are used to structure the operator graph.

> For our purposes it's merely important to know than that they define a hierarchy on the operators.

Reviewer 6: ```suggestion
The dataflow is organized in a hierarchical structure that contains operators and _regions_, which define a hierarchy on the operators.

In our example, the dataflow contains a `InputRegion` and a `BuildRegion`.
And the `BuildRegion` contains the operators from the above plan.

![Regions and operator visualization](/images/regions-and-operators.png)
```

## The system catalog and introspection relations

> Materialize collects a lot of useful information about the dataflows and operators in the System Catalog in [introspection relations](/sql/system-catalog/mz_internal/#replica-introspection-relations).

Reviewer 6: ```suggestion
Materialize collects a lot of useful information about the dataflows and operators in system catalog [introspection relations](/sql/system-catalog/mz_internal/#replica-introspection-relations).
```

The introspection relations are extremely valuable to troubleshoot and to understand what is happening under the hood,

> in particular if Materialize is not behaving in the expected way.

Reviewer 6: ```suggestion
The introspection relations are useful to troubleshoot and understand what is happening under the hood when Materialize is not behaving as expected.
```

However, it is important to understand that most of the statistics we need for troubleshooting purposes are specific to the cluster that is running the queries we want to debug.

> {{< warning >}}
> The contents of introspection relations differ depending on the selected cluster and replica.
> As a consequence, you should expect the answers to the queries below to vary depending on which cluster you are working in.
> In particular, indexes and dataflows are local to a cluster, so their introspection information will vary across clusters.
> {{< /warning >}}

Reviewer 6: ```suggestion
{{< warning >}}
Indexes and dataflows are local to a cluster, so their introspection information will vary across clusters depending on the active cluster and replica.
As a consequence, you should expect the results of the queries below to vary depending on the values set for the `cluster` and `cluster_replica` [session variables](/sql/set/#session-variables).
{{< /warning >}}
```

<!--
[//]: # "TODO(joacoc) We should share ways for the user to diagnose and troubleshoot if and how fast a source is consuming."
``` -->

> ## Where is Materialize spending time on computations?

Reviewer 6: ```suggestion
## Where is Materialize spending compute time?
```

Materialize spends time in various dataflow operators maintaining
materialized views or indexes. If Materialize is taking more time to update
results than you expect, you can identify which operators
take the largest total amount of time.

> ### Finding expensive queries/dataflows

Reviewer 6: ```suggestion
### Identifying expensive dataflows
```

> To understand which query or dataflow, respectively, is taking the most time we can query the `mz_scheduling_elapsed` relation.

Reviewer 6: "Which query or dataflow, respectively". What is this making reference to? `mz_scheduling_elapsed` is bound to a relation, and in the following section you dig into operators per dataflow, so I'm missing the bit where we make references to queries.

Reviewer: Maybe the confusion here comes from unclarity about how SQL constructs map to compute constructs?

* index -> dataflow (with an arrangement at the end)
* materialized view -> dataflow (with a storage sink at the end)
* view -> doesn't touch compute directly
* SELECT query -> either:
  * peek against an arrangement ("fast path") _or_
  * temporary dataflow (with an arrangement at the end) + peek against that arrangement

> The `elapsed_time` metric shows the absolute time the dataflows was busy since the system started and the dataflows was created.

Reviewer 6: ```suggestion
The `elapsed_time` metric shows the absolute time the dataflows was busy since the system started and the dataflow was created.
```

```sql
-- Extract raw elapsed time information for dataflows
SELECT
    mdo.id,
    mdo.name,
    mse.elapsed_ns / 1000 * '1 MICROSECONDS'::interval AS elapsed_time
FROM mz_internal.mz_scheduling_elapsed AS mse,
    mz_internal.mz_dataflow_operators AS mdo,
    mz_internal.mz_dataflow_addresses AS mda
WHERE mse.id = mdo.id AND mdo.id = mda.id AND list_length(address) = 1
ORDER BY elapsed_ns DESC
```
```
 id  |                  name                  |  elapsed_time
-----+----------------------------------------+-----------------
 354 | Dataflow: materialize.qck.num_bids     | 00:11:05.107351
 180 | Dataflow: materialize.qck.avg_bids_idx | 00:09:34.29102
 578 | Dataflow: materialize.qck.num_bids_idx | 00:01:25.363355
(3 rows)
```

These results show that the system spend the most time keeping the materialized view `num_bids` (from the quick start) up to date.

> Followed by the work on the index `avg_bids_idx` that has been defined on a view from the query above.

Reviewer 6: ```suggestion
```

 id  |               name               |             dataflow_name              |  elapsed_time
-----+----------------------------------+----------------------------------------+-----------------
 431 | ArrangeBy[[Column(0)]]           | Dataflow: materialize.qck.num_bids     | 00:06:17.074208
 257 | ArrangeBy[[Column(0)]]           | Dataflow: materialize.qck.avg_bids_idx | 00:06:16.302804
 268 | ArrangeBy[[Column(0)]]           | Dataflow: materialize.qck.avg_bids_idx | 00:00:34.975043
 442 | ArrangeBy[[Column(0)]]           | Dataflow: materialize.qck.num_bids     | 00:00:34.403058
 528 | shard_source_fetch(u517)         | Dataflow: materialize.qck.num_bids     | 00:00:29.972871
 594 | shard_source_fetch(u517)         | Dataflow: materialize.qck.num_bids_idx | 00:00:27.517982
 230 | shard_source_fetch(u510)         | Dataflow: materialize.qck.avg_bids_idx | 00:00:25.494895
 404 | shard_source_fetch(u510)         | Dataflow: materialize.qck.num_bids     | 00:00:22.286794
 566 | persist_sink u517 append_batches | Dataflow: materialize.qck.num_bids     | 00:00:17.657125
 196 | shard_source_fetch(u509)         | Dataflow: materialize.qck.avg_bids_idx | 00:00:08.692289
 ...
```

> ## Why is Materialize unresponsive and where is it currently spending time?

Reviewer 6: ```suggestion
## Why is Materialize unresponsive?
```

            )
    ),

    histograms AS (
        SELECT
            mdo.id,
            mdo.name,
            mdod.dataflow_name,
            mcodh.count,
            mcodh.duration_ns / 1000 * '1 microseconds'::interval AS duration
        FROM mz_internal.mz_compute_operator_durations_histogram AS mcodh,
            mz_internal.mz_dataflow_operators AS mdo,
            mz_internal.mz_dataflow_operator_dataflows AS mdod
        WHERE
            mcodh.id = mdo.id
            AND mdo.id = mdod.id
            AND mdo.id IN (SELECT id FROM leaves)
    )

    SELECT *
    FROM histograms
    WHERE duration > '100 milliseconds'::interval
    ORDER BY dataflow_name ASC, duration DESC
) WITH (SNAPSHOT = false, PROGRESS)) TO STDOUT;
```
```
1691667343000	t	\N	\N	\N	\N	\N	\N
1691667344000	t	\N	\N	\N	\N	\N	\N
1691667344000	f	-1	431	ArrangeBy[[Column(0)]]	Dataflow: materialize.qck.num_bids	7673	00:00:00.104800
1691667344000	f	1	431	ArrangeBy[[Column(0)]]	Dataflow: materialize.qck.num_bids	7674	00:00:00.104800
1691667345000	t	\N	\N	\N	\N	\N	\N...
```

In this way you can see that currently the only operator that is doing more than 100 milliseconds worth of work

> is the `ArrangeBy` operator form the materialized view `num_bids`.

Reviewer 6: ```suggestion
is the `ArrangeBy` operator from the materialized view `num_bids`.
```

as it changes. These queries extract the numbers of records and
batches backing each of the arrangements. The reported records may
exceed the number of logical records; the report reflects the uncompacted
state.

```sql

> -- Extract arrangement records and batches

Reviewer 5: Just a note that in 2 weeks we'll be exposing actual bytes used by arrangement!

Reviewer: This is actually live in prod as of literally now :)
```
materialize=> select * from mz_internal.mz_arrangement_sizes;
 operator_id | records | batches |    size    |  capacity  | allocations
-------------+---------+---------+------------+------------+-------------
 372         |       0 |       1 |         16 |         16 |           7
 770         |      43 |       2 |       4397 |       4397 |          18
 364         |    9179 |       2 |    1028452 |    1028452 |          18
 368         |    9179 |       2 |    1028452 |    1028452 |          18
 755         | 1188533 |       2 |  128877811 |  128877963 |          18
 740         | 1768240 |       2 | 1561397696 | 1561397696 |          18
 802         | 5755315 |       2 | 1391255464 | 1391262526 |          47
 817         | 5755329 |       4 | 1718268926 | 1718280973 |          36
 725         | 5755303 |       3 | 1299892208 | 1299897269 |          18
(9 rows)
```
